File: Case_3.go
package main

import "fmt"

// é usada para calcular o comprimento da maior substring sem caracteres
// repetidos em uma string fornecida. A função recebe uma string s como entrada e retorna
// um número que representa o comprimento da maior substring sem caracteres repetidos.
func lengthOfLongestSubstring(s string) int { 
	charIndexMap := make(map[rune]int) // Cria um map para armazenar o índice de cada caractere
	maxTamanho := 0 									 // Inicializa o tamanho máximo da substring como 0
	inicio := 0 											 // Inicializa o índice de início da substring como 0

	for ate, char := range s { // Percorre a string
		if index, ok := charIndexMap[char]; ok && index >= inicio { // Verifica se o caractere já foi encontrado
			inicio = index + 1 // Atualiza o índice de início da substring
		}
		charIndexMap[char] = ate // Atualiza o índice do caractere
		if ate-inicio+1 > maxTamanho {	// Verifica se o tamanho da substring é maior que o tamanho máximo
			maxTamanho = ate - inicio + 1 // Atualiza o tamanho máximo da substring
		}
	}

	return maxTamanho // Retorna o tamanho máximo da substring
}

func main() {
	fmt.Println(lengthOfLongestSubstring("abcabcbb")) // Output: 3
	fmt.Println(lengthOfLongestSubstring("pwwkew"))   // Output: 3
}

File: Case_3.ts
/**
 * Calcula o comprimento da substring mais longa sem repetir caracteres em uma determinada string.
 * @param s - O input e tipo string.
 * @returns O comprimento da substring mais longa sem caracteres repetidos.
 *
 * lengthOfLongestSubstring()
 * é usada para calcular o comprimento da maior substring sem caracteres
 * repetidos em uma string fornecida. A função recebe uma string s como entrada e retorna
 * um número que representa o comprimento da maior substring sem caracteres repetidos.
 */
function lengthOfLongestSubstring(s: string): number {
  let maxTamanho = 0; // Comprimento da substring mais longa sem caracteres repetidos
  let inicio = 0; // Início da substring
  const charIndexMap: { [key: string]: number } = {}; // Mapeia o caractere para o seu índice

  for (let ate = 0; ate < s.length; ate++) {
    // Loop através da string
    const char = s[ate]; // Caractere atual
    if (charIndexMap[char] !== undefined) {
      // Se o caractere já existe no mapa
      inicio = Math.max(inicio, charIndexMap[char] + 1); // Atualiza o início da substring
    }
    charIndexMap[char] = ate; // Atualiza o índice do caractere
    maxTamanho = Math.max(maxTamanho, ate - inicio + 1); // Atualiza o comprimento da substring
  }

  return maxTamanho;
}

// Exemplos de uso
console.log(lengthOfLongestSubstring("abcabcbb")); // 3
console.log(lengthOfLongestSubstring("pwwkew")); // 3
